# Sandbox 1

April 22, 2021

## 1. How to convert list of tuples into np array

In [2]:
import numpy as np

In [20]:
li = []
for i in range(10, 16):
    tup = (i, i**2)
    li.append(tup)
    
li

[(10, 100), (11, 121), (12, 144), (13, 169), (14, 196), (15, 225)]

In [33]:
arr = np.array(li)

arr[:,0]

array([10, 11, 12, 13, 14, 15])

In [22]:
arr[:,1]

array([100, 121, 144, 169, 196, 225])

#### That was easy

## 2. Index slicing on np arrays

In [23]:
arr

array([[ 10, 100],
       [ 11, 121],
       [ 12, 144],
       [ 13, 169],
       [ 14, 196],
       [ 15, 225]])

#### Get the min and max x, y values

In [13]:
x_min = arr[0,0]

x_min

10

In [14]:
y_max = arr[-1,1]

y_max

225

## 3. Make a deep copy of np array

#### First, a shallow copy

In [24]:
arr

array([[ 10, 100],
       [ 11, 121],
       [ 12, 144],
       [ 13, 169],
       [ 14, 196],
       [ 15, 225]])

In [34]:
arr2 = arr

In [35]:
arr[:,1] = arr[:,1]-99

In [36]:
arr

array([[ 10,   1],
       [ 11,  22],
       [ 12,  45],
       [ 13,  70],
       [ 14,  97],
       [ 15, 126]])

In [38]:
arr2

array([[ 10,   1],
       [ 11,  22],
       [ 12,  45],
       [ 13,  70],
       [ 14,  97],
       [ 15, 126]])

#### Deep copy

In [40]:
arr3 = arr.copy()

In [41]:
arr3

array([[ 10,   1],
       [ 11,  22],
       [ 12,  45],
       [ 13,  70],
       [ 14,  97],
       [ 15, 126]])

In [42]:
arr[:,1] = arr[:,1] + 99

arr

array([[ 10, 100],
       [ 11, 121],
       [ 12, 144],
       [ 13, 169],
       [ 14, 196],
       [ 15, 225]])

In [43]:
arr3

array([[ 10,   1],
       [ 11,  22],
       [ 12,  45],
       [ 13,  70],
       [ 14,  97],
       [ 15, 126]])

## 3. Inspect dataset

In [3]:
import pandas as pd

In [46]:
arr = np.load('../data/arr50.npy')

In [52]:
df = pd.DataFrame(arr, columns = ['fare', 'age', 'survived'])

In [54]:
df.describe()

fare        age   survived
count   50.000000  50.000000  50.000000
mean    30.334498  30.190000   0.340000
std     31.164935  18.385578   0.478518
min      0.000000   2.000000   0.000000
25%      8.801050  17.000000   0.000000
50%     20.400000  28.000000   0.000000
75%     31.521850  42.000000   1.000000
max    120.000000  71.000000   1.000000

In [57]:
df[df.fare > 0].min()

fare        6.975
age         2.000
survived    0.000
dtype: float64

#### Note: 6.975 was 817 GBP, or 1400 CAD

## Vector conversions
I'm currently using iteration to normalize raw Titanic data

In [5]:
df = pd.read_csv('data/titanic.csv')

In [7]:
df = df[['fare', 'age', 'survived']]

In [8]:
df.head()

fare   age  survived
0   7.2500  22.0         0
1  71.2833  38.0         1
2   7.9250  26.0         1
3  53.1000  35.0         1
4   8.0500  35.0         0

In [12]:
df = df.dropna()

In [13]:
df.describe()

fare         age    survived
count  714.000000  714.000000  714.000000
mean    34.694514   29.699118    0.406162
std     52.918930   14.526497    0.491460
min      0.000000    0.420000    0.000000
25%      8.050000   20.125000    0.000000
50%     15.741700   28.000000    0.000000
75%     33.375000   38.000000    1.000000
max    512.329200   80.000000    1.000000

## Add col for tracking nn
- 0 = not a nearest neightbour

In [19]:
df['nn'] = 0

df.head()

fare   age  survived  nn
0   7.2500  22.0         0   0
1  71.2833  38.0         1   0
2   7.9250  26.0         1   0
3  53.1000  35.0         1   0
4   8.0500  35.0         0   0

### Convert to numpy

In [14]:
arr = df.to_numpy()
len(arr)

714

In [16]:
arr.shape

(714, 3)

In [18]:
arr[:,3] = 0

IndexError: index 3 is out of bounds for axis 1 with size 3

## Add col to numpy array

<img src='https://www.w3resource.com/w3r_images/python-numpy-image-exercise-86.png' width = 200>

In [20]:
import numpy as np

### Simple example

In [53]:
x = np.array([[10,20,30], [40,50,60]])

y = np.array([[100], [200]])

x.shape

(2, 3)

In [54]:
y.shape

(2, 1)

In [57]:
z = np.append(x, y, axis=1)

z

array([[ 10,  20,  30, 100],
       [ 40,  50,  60, 200]])

In [58]:
z.shape

(2, 4)

### Example 2

#### Make 1d arr

In [62]:
shape = (10, 2)

arr = np.zeros(shape)
arr

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [47]:
arr.shape

(10, 2)

#### Make another 1d arr

In [67]:
arr2 = np.zeros((10, 1))
arr2

array([[0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.]])

In [68]:
arr2.shape

(10, 1)

#### Append

In [69]:
np.append(arr, arr2, axis=1)

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

### With titanic data

In [81]:
arr = np.load('../data/arr.npy')

In [82]:
arr.shape

(714, 3)

In [83]:
arr2 = np.zeros( (len(arr), 2) )
arr2.shape

(714, 2)

In [84]:
arr = np.append(arr, arr2, axis = 1)

In [85]:
arr.shape

(714, 5)

In [86]:
arr[:5]

array([[ 7.25  , 22.    ,  0.    ,  0.    ,  0.    ],
       [71.2833, 38.    ,  1.    ,  0.    ,  0.    ],
       [ 7.925 , 26.    ,  1.    ,  0.    ,  0.    ],
       [53.1   , 35.    ,  1.    ,  0.    ,  0.    ],
       [ 8.05  , 35.    ,  0.    ,  0.    ,  0.    ]])

In [99]:
arr[:,:-2]

array([[ 7.25  , 22.    ,  0.    ],
       [71.2833, 38.    ,  1.    ],
       [ 7.925 , 26.    ,  1.    ],
       ...,
       [30.    , 19.    ,  1.    ],
       [30.    , 26.    ,  1.    ],
       [ 7.75  , 32.    ,  0.    ]])

## Merge my linspace arrays into one arr

In [100]:
x = np.linspace(0, 20, 7)
y = np.linspace(0, 100, 7)

In [104]:
x.shape

(7,)

In [105]:
y.shape

(7,)

In [125]:
x = x.reshape(7,1)
y = y.reshape(7,1)

In [129]:
np.concatenate((x, y), axis=1)

array([[  0.        ,   0.        ],
       [  3.33333333,  16.66666667],
       [  6.66666667,  33.33333333],
       [ 10.        ,  50.        ],
       [ 13.33333333,  66.66666667],
       [ 16.66666667,  83.33333333],
       [ 20.        , 100.        ]])

In [127]:
z.shape

(14, 1)

In [120]:
x.shape

(7,)

In [121]:
x.reshape(7,1)

array([[ 0.        ],
       [ 3.33333333],
       [ 6.66666667],
       [10.        ],
       [13.33333333],
       [16.66666667],
       [20.        ]])

In [123]:
x.shape

(7,)

In [131]:
z = np.linspace(1,10,5).reshape(5,1)

In [132]:
z.shape

(5, 1)